File: elixir/lib/aoc2025/secret_entrance.ex
defmodule Aoc2025.SecretEntrance do
  def secret_entrance data do
    Enum.map(data, fn(lyne) ->
      case String.at(lyne, 0) do
        "L" ->  -String.to_integer(String.slice(lyne, 1..-1//1))
        "R" ->  String.to_integer(String.slice(lyne, 1..-1//1))
      end
    end)
  end

  def zeroes turns do
    {_, combination} = Enum.reduce(turns, {50, 0}, fn(turn, memo) ->  # reverse of ruby binding
      {position, combination} = memo

      # 100+ because rem(-18, 100) yields -18
      position = rem(100+position+turn, 100)
      if (position == 0) do
        {position, combination+1}
      else
        {position, combination}
      end
    end)

    combination
  end

  def clicks turns do
    {_, combination} = Enum.reduce(turns, {50, 0}, fn(turn, memo) ->
      {position, combination} = memo
      combination =
        if turn < 0 do
          combination - div(turn, 100)
        else
          combination + div(turn, 100)
        end
      landing = position + rem(turn, 100)

      cond do
        landing < 0 ->
          if position == 0 do
            {landing+100, combination}
          else
            {landing+100, combination+1}
          end
        landing == 0 ->
          {landing, combination+1}
        landing < 100 ->
          {landing, combination}
        true ->
          {landing-100, combination+1}
      end
    end)

    combination
  end
end

File: go/lib/aoc2025/secret_entrance.go
package aoc2025

import (
	"strconv"
)

type SecretEntrance struct {
	Turns []int
}


/*	Seems that idiomatic go is to have this New$(STRUCTNAME) constructor
	function that takes in the dependent data and creates the structure
	around whatever you do to said data, and returns a reference to same.
*/
func NewSecretEntrance(turns []string) *SecretEntrance {
	me := &SecretEntrance {
		Turns: make([]int, 0),
	}

	//	encode left turns as negative
	for _, turn := range turns {
		amount, _ := strconv.Atoi(turn[1:])
		switch turn[0] {
		case 'L':
			me.Turns = append(me.Turns, amount * -1)
		case 'R':
			me.Turns = append(me.Turns, amount)
		default:
			//
		}
	}

	return me
}


func (me SecretEntrance) Zeroes() int {
	combination := 0
	position := 50

	for _, turn := range me.Turns {
		position += turn
		position %= 100
		if position == 0 {
			combination++
		}
	}

	return combination
}


func (me SecretEntrance) Clicks() int {
	combination := 0
	position := 50
	landing := 0

	for _, turn := range me.Turns {
		if turn < 0 {
			combination -= turn / 100
		} else {
			combination += turn / 100
		}
		landing = position + (turn % 100)

		switch {
		case landing < 0:
			if position != 0 {
				combination++
			}
			position = landing + 100
		case landing == 0:
			combination++
			position = landing
		case landing < 100:
			position = landing
		default:
			combination++
			position = landing - 100
		}
	}

	return combination
}

